\chapter[A historical perspective on bioinformatics]%
{A historical perspective on bioinformatics}
Many decades ago, around the 1950's, we did not know much about the molecules that carry our genetic information and how this would be translated into higher levels of biology. All what was known about deoxyribonucleic acid (DNA) was that it carries nucleotides in equimolar proportions, so that there is as much guanine as cytodine and as much adenine as thymine. A major breakthrough came with Watson's and Crick's discovery of the structure of DNA in 1953\cite{dnastruct}. Despite that, it took some more decades before the genetic code was deciphered and how this information is further transferred. Researches came to know that DNA is essentially build up of a linear sequence of the 4 aforementioned nucleic acids. This sequence encodes information that undergoes transcription into ribonucleic acid (RNA) that in turn gets translated into proteins. The encoding of proteins is done in groups of three, known as codons. All of this was later stated as the \textit{centra dogma of molecular biology}, also by Crick in 1958\cite{dogma}. This was hugely important for later research since it gives us more insight on how the genetic code is translated and transferred.

In the same decade, major leaps were made in the research of protein structure and sequences. The first three-dimensional protein structures were determined via X-ray crystallography \cite{xray}, which is still mostly the preferred method to this day. On top of that, the arrangement of the primary structure of a protein has been resolved after the first sequencing of a polypeptide. Sanger determined by sequencing insulin in 1953\cite{insulin} that a protein is build up of a sequence of amino acids, all connected by a peptide bond into a polypeptide. This established the idea that proteins are biological macromolecules that carried lots of information \cite{primstruct} and so there came a boom of research on more efficient methods for obtaining protein sequences. The popular method of that time was the Edman degradation method. The sequencing of amino acids was also rapidly made automated later in the 1960's. A major issue with this method was that only a theoretical maximum of 50 to 60 sequential amino acids could be sequenced. Larger proteins had to be cleaved into fragements that were small enough to be sequenced. To trace back the input sequence was a cumbersome process and thus 