\section{A historical perspective on bioinformatics}
Many decades ago, around the 1950s, we did not know much about the molecules that carry our genetic information and how this would be translated into higher levels of biology. All that was known about deoxyribonucleic acid (DNA) was that it carries nucleotides in equimolar proportions so that there is as much guanine as cytodine and as much adenine as thymine. A major breakthrough came with J. Watson, F. Crick and R. Franklin's discovery of the structure of DNA in 1953~\cite{dnastruct}~\cite{franklin}. Despite that, it took several more decades before the genetic code was deciphered and how this information is further processed. Researchers came to know that DNA is essentially built up of a linear sequence of the four aforementioned nucleic acids. This sequence encodes information that undergoes transcription into ribonucleic acid (RNA) which in turn gets translated into proteins. The encoding of proteins is done in groups of three nucleic acids at a time, known as codons. This was later summarized as the \textit{central dogma of molecular biology}, also by Crick in 1958~\cite{dogma}. This was hugely important for later research since it gives us more insight into how the genetic code is translated and transferred.

\begin{figure}[h]
    \centering
    \vspace{-80pt}
    \startchronology[startyear=1948,stopyear=2018, height=0.5ex, dates=false]
    \chronoevent[textwidth=2.5cm,markdepth=-90pt,flipped]{1953}{\parbox{2.5cm}{\scriptsize\raggedright{Double helix model for DNA~\cite{dnastruct}. Insulin sequenced by Sagner~\cite{insulin}}}}
    \chronoevent[textwidth=2.5cm,markdepth=90pt]{1958}{\parbox{2.5cm}{\scriptsize\raggedright{Central Dogma proposed by Crick~\cite{dogma}}}}
    \chronoevent[textwidth=2.5cm,markdepth=-30pt,flipped]{1962}{\parbox{2.5cm}{\scriptsize\raggedright{COMPROTEIN by Dayhoff~\cite{comprotein}}}}
    \chronoevent[textwidth=2.5cm,markdepth=30pt]{1965}{\parbox{2.5cm}{\scriptsize\raggedright{First protein database by Dayhoff~\cite{atlas}}}}
    \chronoevent[textwidth=2.5cm,markdepth=-90pt,flipped
    ]{1968}{\parbox{2.5cm}{\scriptsize\raggedright{Genetic code deciphered~\cite{codon}}}}
    \chronoevent[textwidth=2.5cm,markdepth=90pt]{1970}{\parbox{2.5cm}{\scriptsize\raggedright{Needleman \& Munsch's global alignment algorithm~\cite{global}}}}
    \chronoevent[textwidth=2.5cm,markdepth=-30pt,flipped]{1977}{\parbox{2.5cm}{\scriptsize\raggedright{First full genome sequenced (Phi X174)~\cite{dnaseq}}}}
    \chronoevent[textwidth=2.5cm,markdepth=30pt]{1981}{\parbox{2.5cm}{\scriptsize\raggedright{Smith \& Waterman's local alignment algorithm~\cite{local}}}}
    \chronoevent[textwidth=2.5cm,markdepth=-90pt,flipped]{1987}{\parbox{2.5cm}{\scriptsize\raggedright{FASTA for DNA and protein searches released~\cite{fasta}}}}
    \chronoevent[textwidth=2.5cm,markdepth=90pt]{1991}{\parbox{2.5cm}{\scriptsize\raggedright{Human Genome project initiated~\cite{hgpc}}}}
    \chronoevent[textwidth=2.5cm,markdepth=-30pt,flipped]{2003}{\parbox{2.5cm}{\scriptsize\raggedright{Human genome fully mapped~\cite{hgpc}}}}
    \chronoevent[textwidth=2.5cm,markdepth=30pt]{2005}{\parbox{2.5cm}{\scriptsize\raggedright{Next-gen sequencing technology publicly available~\cite{ngs}}}}
    \chronoevent[textwidth=2.5cm,markdepth=-90pt,flipped]{2018}{\parbox{2.5cm}{\scriptsize\raggedright{AlphaFold introduced~\cite{alphafold}}}}
    \stopchronology
    \caption{Timeline of important developments in bioinformatics.}
    \label{fig:timeline}
\end{figure}


In the same decade, major leaps were made in the research of protein structure and sequences. The first three-dimensional protein structures were determined via X-ray crystallography~\cite{xray}, which is still to this day the preferred method. On top of that, the arrangement of the primary structure of a protein has been resolved after the first sequencing of a polypeptide. Sanger determined by sequencing insulin in 1953 that a protein is built up of a sequence of amino acids, all connected by a peptide bond to form a polypeptide~\cite{insulin}. This finding established the idea that proteins are biological macromolecules that carry lots of information~\cite{primstruct}, starting a boom of research on more efficient methods for obtaining protein sequences. The most popular method of that time was the Edman degradation method~\cite{edman}. A major issue with this method was that only a theoretical maximum of 50 to 60 sequential amino acids could be sequenced. Larger proteins had to be cleaved into fragments that were small enough to be sequenced. Tracing back the input sequence from this data was a cumbersome process and thus published Dayhoff the first computational program applied to biological data, COMPROTEIN~\cite{comprotein} in 1962. This program was essentially a \mbox{\textit{de novo}} sequence assembler for Edman degradation data. Furthermore, the sequencing of amino acids was also increasingly made automated later in the 1960s. These innovations assisted the creation of the first published protein sequence database~\cite{atlas} in 1965. 

Following Figure~\ref{fig:timeline} further into the 1960s, researchers discovered the evolutionary value of having protein sequence data of different species. The problem to be solved back then was the quantification of the similarity between sequences. Pairwise alignment algorithms such as the algorithm of Needleman and Wunsch for global alignments~\cite{global} and that of Smith and Waterman for local alignments~\cite{local} from the 1970s solved this issue and alignment is still considered to be a key bioinformatics task to this day. Together with alignments, mathematical frameworks for amino acid substitutions in the context of evolution such as PAMs and BLOSUMs also contributed to bioinformatics. These pairwise alignment algorithms are sufficient for comparing two sequences but unfeasible for searching databases for homologous sequences, hence faster algorithms like FASTA~\cite{fasta} and Basic Local Alignment Search Tool (BLAST)~\cite{blast} were developed in the 1980s and 1990s. These methods were and are still important for discovering functional, structural and evolutionary information in biological sequences since sequences demonstrating homology, or similarity due to shared ancestry, are highly likely to possess the same biological function. This also means that such sequences might be derived from a common ancestor and it became commonplace that sequence patterns may lead to structural and functional relevance. A natural extension of pairwise alignments is multiple sequence alignment (MSA)~\cite{msa}, which concerns aligning multiple related sequences. The most popular and still widely used tools today include Clustal~\cite{clustal} and MUSCLE~\cite{muscle}. MSAs reveal much more information than pairwise alignment can. They allow for the identification of conserved sequence patterns and critical amino acid residues with much more statistical significance which is of great value for constructing phylogenetic profiles of gene families~\cite{phylo}. All of this showed the importance of computational biology and established bioinformatics as a beneficial field of science.

Development of DNA-based applications took some more time since the genetic code and how it translates to amino acids was not deciphered yet until 1968~\cite{codon}. Early DNA and RNA sequencing methods were first demonstrated in the 1970s~\cite{dnaseq, rna} and like with protein sequencing methods, these methods only became faster, more efficient and more scalable. The 1990s saw the appearance of whole genome sequencing and internet-accessible databases that we still use to this day, such as Genbank~\cite{genbank}, Ensembl~\cite{ensembl} and the European Nucleotide Archive (ENA)~\cite{ena}. This led to significant advancements in computational biology, driven by the continuous increase in data volume and the escalating demand for computational processing power. With the advent of second-generation sequencing in the 2000s came even more Big Data issues, further challenging bioinformaticians by allowing us to sequence millions of DNA and RNA base pairs in a single run. The sheer size and complexity of biological data can make it difficult to store and manage, as for instance implementing error identification, security, quality standards and easy data retrievability. On top of that, the analysis of such large-scale data is not straightforward and traditional software tools won't be sufficient anymore. Lastly, the need for high-performance computing resources to handle all of these computational demands will only rise.

Today we see that research in the field of biology is becoming more and more computation-driven and that this trend will not slow down any time soon~\cite{compbio}. This thesis will specifically concentrate on the role of computational approaches in protein research, which is detailed in the subsequent section.